### PR TITLE
Histogram Lua class

### PR DESCRIPTION
**Pull request by cody271**

### CODE REVIEW

**Comment by HeinrichHartmann** [comment]
Still digging though, but AFAIK, this PR looks correct and provides a drop-in replacement for our Histogram_v1.lua module that avoids table allocations.

However, I question if this is the right place for a module like this.

An official lua API wrapper would ideally be minimalistic and just wrapping the C functions, like our Python wrapper: https://github.com/circonus-labs/libcircllhist/blob/master/src/python/circllhist/wrapper.py
I would like to see us doing something similar for lua.

Differences to this version are as follows:

1. This module pulls in 3 different dependencies in that are not needed in a minimal version: funiter (worst offender), tablex, lunity. This feels bloaty and duplicates code.

2. In addition, Histogram_v1 contains a number of regrettable choices that we want to get away from.
By exposing these functions here, we morally oblige ourself to keep on supporting them.
- coroutines for IV quantiles are really slow https://github.com/circonus-labs/libcircllhist/pull/38/files#diff-a46588a3039863238263e1d9739d3175R243
- auto_y_rescale is off-topic for libcircllhist https://github.com/circonus-labs/libcircllhist/pull/38/files#diff-a46588a3039863238263e1d9739d3175R334
- self:quantiles(). We create a table just to unpack it. Wasteful. https://github.com/circonus-labs/libcircllhist/pull/38/files#diff-a46588a3039863238263e1d9739d3175R217
- Same here: https://github.com/circonus-labs/libcircllhist/pull/38/files#diff-a46588a3039863238263e1d9739d3175R226
- Same here: https://github.com/circonus-labs/libcircllhist/pull/38/files#diff-a46588a3039863238263e1d9739d3175R287
- Same here: https://github.com/circonus-labs/libcircllhist/pull/38/files#diff-a46588a3039863238263e1d9739d3175R302
- Same here: https://github.com/circonus-labs/libcircllhist/pull/38/files#diff-a46588a3039863238263e1d9739d3175R316
- coros for Hist:iteration again. This is a killer: https://github.com/circonus-labs/libcircllhist/pull/38/files#diff-a46588a3039863238263e1d9739d3175R440
- Why expose Histogram_C and Histogram as names for classes (unmotivated here)

So my suggestion would be to either:
(a) Move this module into our internal code base as Histogram_v2.lua, and strip out the dependencies mentioned in (1). OR
(b) Put a big disclaimer at the top, that the exposed API is unstable and will undergo changes as we cleanup the Histogram module.

File: src/lua/circllhist.lua [hunk 1]
```diff
@@ -0,0 +1,565 @@
...
+  if idx > param.limit then
+    return nil -- end iteration
+  end
+  local val, count = param.self:bucket_idx(idx)
+  return idx + 1, val, count
+end
+local function histogram_filter_count_zero(val, count)
+  return not (count == 0)
+end
+function Histogram_C:funiter()
```
Comment (HeinrichHartmann) [inline src/lua/circllhist.lua]: This is better than the coro-iteration, but depends on the fun.lua.

Iterating over the histogram with:
```
for i=1, h:bucket_count() do
  val, count = h:bucket_idx(idx)
  ...
end
```
should be good enough, no?

Consumers are free to wrap it into a fun iterator themselves.

File: src/lua/circllhist.lua [hunk 2]
```diff
@@ -0,0 +1,565 @@
...
+    __pairs=function() return pairs(t) end,
+    __ipairs=function() return ipairs(t) end,
+    __len=function() return #t end,
+    __metatable=false
+  }
+  return setmetatable({}, mt)
+end
+
+-- fun.lua, MIT/X11
+local fun = {}
```
Comment (HeinrichHartmann) [inline src/lua/circllhist.lua]: fun.lua is only used to provide convenient iteration methods. No _need_ to have this here. We provide other means for iteration (cf. below)

File: src/lua/circllhist.lua [hunk 3]
```diff
@@ -0,0 +1,565 @@
...
+  local total = 0 -- count non nans
+  for _, bucket_bound, count in self:funiter():filter(vc_not_nan) do
+    total = total + count
+  end
+
+  if total == 0 then
+    return {} -- no values for empty histogram
+  end
+
+  local ivq_iter = coroutine.wrap(function(val)
```
Comment (HeinrichHartmann) [inline src/lua/circllhist.lua]: Use of coroutines is expensive. Avoid here.

File: src/lua/circllhist.lua [hunk 4]
```diff
@@ -0,0 +1,565 @@
...
+end
+
+--- returns the number of values in a histogram that are in buckets
+--- that are entirely lower or equal than a given value.
+function Histogram_C:count_below(...)
+  local out = {}
+  for i = 1, select("#", ...) do
+    out[i] = self:_count_below_single(select(i, ...))
+  end
+  return unpack(out)
```
Comment (HeinrichHartmann) [inline src/lua/circllhist.lua]: We create a table just to unpack it below. Is much faster to do this via recursion.

Also why do we want to support a multi argument `count_below()` in the API? We only have a single argument version in C. This is a pure convenience feature.

Comment (HeinrichHartmann) [reply src/lua/circllhist.lua]: Similar arguments apply to count_above, count_nearby, and others.

File: src/lua/circllhist.lua [hunk 5]
```diff
@@ -0,0 +1,565 @@
...
+function Histogram_C:merge(other_histogram)
+  for _, b, c in other_histogram:funiter() do
+    self:insert(b,c);
+  end
+end
+
+function Histogram_C:merge_c(other_histogram_c)
+  local other_histo_ptr = ffi.new("const histogram_t*[1]", other_histogram_c)
+  libhist.hist_accumulate(self, other_histo_ptr, 1);
+end
```
Comment (HeinrichHartmann) [inline src/lua/circllhist.lua]: Why support two merge methods? This makes sense if we have multiple implementations for the same interface (which we had in Histogram_v1). There is only one here, so this is unmotivated.

File: src/lua/circllhist.lua [hunk 6]
```diff
@@ -0,0 +1,565 @@
...
+    return ffi.string(ffi_bucket_str_buffer, bucket_str_len)
+  elseif format == "hex" then
+    return '0x' .. bit.tohex(bit.bor(ffi_bucket.exp, bit.lshift(ffi_bucket.val,8)), 4)
+  else
+    return string.format(format, libhist.hist_bucket_to_double(ffi_bucket))
+  end
+end
+-- fast, standartized JSON encoder
+-- This function will be used by dkjson to serialize Histogam_C objects
+function Histogram_C:__tojson(state)
```
Comment (HeinrichHartmann) [inline src/lua/circllhist.lua]: Just call this `:tojson()` like we have `:tostring()` or `:totable()`. Users might not care about dkjson.

File: src/lua/circllhist.lua [hunk 7]
```diff
@@ -0,0 +1,565 @@
...
+end
+
+-- old name
+Histogram_C.count_bucket = Histogram_C.count_nearby
+
+--- Determine sensible limits for the y-axis
+-- Method 0: Calculate global min + max, round to next bucket.
+-- Method 1: If we have little data use Method 0.
+--   Otherwise: Arrange graph so that 75% percentile should be at a comfortable position
+function Histogram_C:auto_y_scale()
```
[the remark this answers is not in the document]
Comment (rileyberton) [reply src/lua/circllhist.lua]: I don't think there is danger in adding this to libcircllhist.  There is nothing prioprietary in adding some math to determine acceptable Y axis.

**Comment by rileyberton** [comment]
I disagree that we need to make this module pristine before releasing.  This is open source free software that is very useful to people who do Lua development.  I would release as is and fix over time.

**Comment by HeinrichHartmann** [comment]
> I disagree that we need to make this module pristine before releasing.

I never said this. I just want to be clear, that I have no intentions in supporting a module like this going forward. Breaking changes will land without warning.

File: src/lua/circllhist.lua (hunk 7 above)
Comment (rileyberton) [inline src/lua/circllhist.lua]: This needs come out and become a static function inside the snowth code base.

File: src/lua/circllhist.lua [hunk 8]
```diff
@@ -0,0 +1,565 @@
...
+-- returns ratios of samples that are lower than the provided values.
+-- E.g.
+-- a, b = inverse_quantiles(10, 15)
+-- a = ratio of samples <10
+-- b = ratio of samples <15
+function Histogram_C:inverse_quantiles(...)
+  return unpack(self:_inverse_quantiles({...}))
+end
+
+function Histogram_C:_inverse_quantiles(thresholds)
```
Comment (rileyberton) [inline src/lua/circllhist.lua]: This needs to come out and move to a static function inside snowth

**Pull request closed without merging**